### PR TITLE
feat(remap): added `parse_key_value` remap function

### LINKED ISSUE #5347: transforms[key_value_parser]: when values contains separator character(s) ,value occur data loss
### Vector Configuration File

```toml
[sources.syslog_514]
  type = "syslog"
  address = "0.0.0.0:514"
  mode = "udp"

[transforms.source_ip]
  type = "split"
  inputs = ["syslog_514"]
  drop_field = false
  field = "source_ip"
  field_names = ["dev_ip"]
  separator = ":"

[transforms.kv_parser_transform]
  type = "key_value_parser"
  inputs = ["source_ip"]
  drop_field = false
  field = "message"
  field_split = "="
  #separator = " "
  #target_field = "data"
  #trim_key = "\""
  trim_value = "\""

[sinks.syslog_514_out]
  inputs   = ["kv_parser_transform"]
  type     = "console"
  encoding = "json"
```

### Debug Sink  Output

{
    "Action": "PERMIT",
    **"Content": "Session",**
    "DstIP": "10.10.254.7",
    "DstPort": "137",
    **"GenTime": "2019-08-19"**,
    "PolicyID": "1",
    "Protocol": "UDP",
    "SerialNum": "100018002000001906146520",
    "SrcIP": "10.10.254.2",
    "SrcPort": "137",
    "appname": "FILTER",
    "facility": "local5",
    "host": "QSFW-1",
    "hostname": "QSFW-1",
    "message": "SerialNum=100018002000001906146520 **GenTime=\"2019-08-19 15:17:44\"** SrcIP=10.10.254.2 DstIP=10.10.254.7 Protocol=UDP SrcPort=137 DstPort=137 PolicyID=1 Action=PERMIT **Content=\"Session Backout\"**",
    "severity": "alert",
    "source_ip": "10.65.60.180",
    "source_type": "syslog",
    "timestamp": "2020-08-19T07:17:44Z"
}

### Expected Behavior
json data key_value
   "GenTime":"2019-10-24 14:25:03"
    "Content": "Session Backout"

### Actual Behavior
json data key_value
   "GenTime":"2019-10-24"
   "Content": "Session"

### Example Data

<318>Oct 24 14:25:03 QSFW-1 FILTER: SerialNum=100018002000001906146520 GenTime="2019-10-24 14:25:03" SrcIP=10.10.254.2 DstIP=10.10.254.7 Protocol=UDP SrcPort=137 DstPort=137 PolicyID=3 Action=PERMIT Content="Session Backout"
<710>Aug 19 16:04:52 QSFW-1 APPCONTROL: SerialNum=100018002000001906146520 GenTime="2019-08-19 16:04:52" SrcIP=10.10.254.2 DstIP=10.10.254.7 Protocol=UDP SrcPort=137 DstPort=137 InInterface=ge0/2 AAA=any App=netbios-ns Action=PERMIT PolicyID=1 AppAction=any Content="The packet is allowed"

### Additional Context

 Vector running in windows and centos 7.3

download package 
   vector-0.11.0-x86_64-pc-windows-msvc
   vector-0.11.0-x86_64-unknown-linux-musl

### LINKED ISSUE #5368: New `parse_key_value` Remap function
To replace the [`key_value_parser` transform](https://vector.dev/docs/reference/transforms/key_value_parser/) and [`logfmt_parser` transform](https://vector.dev/docs/reference/transforms/logfmt_parser/). This was designed after the [Logstash kv parser](https://www.elastic.co/guide/en/logstash/current/plugins-filters-kv.html).

### PR DESCRIPTION
**Pull request by FungusHumungus**
Closes #5368 
Closes #5347 

Adds a `parse_key_value` remap function.

With the default options it will parse key value strings in the format: `name=value name2=value2`.

If there are no parseable fields eg.. `parse_key_value("something")` it just returns an empty map. I have decided to avoid erroring here to keep things simple and infallible, let me know if you think otherwise..

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
Nice work! In general this looks good, but #5368 mentions this should replace the `logfmt` parser too so I think there is one missing feature: allowing values and keys to be wrapped in a delimiter, like `"`s, if they contain the separator character.

For example, this logfmt line: `level=info msg="Stopping all fetchers" tag=stopping_fetchers id=ConsumerFetcherManager-1382721708341 module=kafka.consumer.ConsumerFetcherManager'` parses as `{"id":"ConsumerFetcherManager-1382721708341","level":"info","module":"kafka.consumer.ConsumerFetcherManager","msg":"\"Stopping","tag":"stopping_fetchers"}`.

The example line is from https://www.brandur.org/logfmt

This is a pre-existing issue with the `key_value_parser` transform though, https://github.com/timberio/vector/issues/5347, so I'm ok with opening a separate issue to cover it here.

**Comment by binarylogic** [comment]
Yes, in general, I would make sure to cover the same test cases for both the `key_value_parser` and the `logfmt_parser`. I think we should try to solve both in this PR, if possible.

**Comment by FungusHumungus** [comment]
>  I think there is one missing feature: allowing values and keys to be wrapped in a delimiter, like `"`s, if they contain the separator character.

Oh, yes. Good call.

**Comment by binarylogic** [comment]
> If there are no parseable fields eg.. parse_key_value("something") it just returns an empty map. I have decided to avoid erroring here to keep things simple and infallible, let me know if you think otherwise.

We should error if the parsing fails 😄, like our other `parse_` functions.

**Comment by FungusHumungus** [comment]
With the key_value_parser transform it is possible to specify a split, a separator and characters to trim around the key and value. These trim characters would be the delimiters used to specify the field. Since it is possible to specify multiple characters to trim, I imagine the aim is to be able to parse lines like:

```
<key1>:[value1] <key2>:[value2]
```

where `trim_key = "<>"` and `trim_value = "[]"`

But because the transform is fairly simplistic, it also would work with this, admittedly ridiculous, line:

```
<key1<:]]value1] >key2<:]value2[
```

By updating the parser to be able to handle delimited fields, we need to be clear what the delimiters are. I'm not 100% sure what the best way to specify these would be.

Say the user specified a delimiter of: `"<>[]"`. Clearly, this would be to handle a field like `<a field>` and a field like `[a field]`.

 - Would the user need to be able to specify multiple delimiter characters within a single line?

If so, I suggest we allow the user to specify an array of delimiters, each set has to be either one or two characters long to specify the start and the end - eg. `parse_key_value(.foo, value_delim = ["<>", "[]", "\""])`.

 - Is there a need for a user to have a multi character delimeter? 

For example, would the need a field like `<(field)>` to extract `field`? That would have been possible with the old transform, but not possible with this function.

In summary, I think by fixing the function to avoid splitting within the delimiters, we will be removing some of the flexibility around defining these delimiters. But, I'm not sure if that is really a problem.

**Comment by binarylogic** [comment]
@FungusHumungus I was not aware of that. I thought they were mostly solving the same problem. To back up, we just need something that generally parses key/value strings. We can add more advanced options like trim characters and delimiters later.

If we can cover the following formats that should be a good start.

Sematics:

* Space delimited: `key1=value1 key2=value2`
* Quoted values: `key="with special=characters"`
* Quoted keys: `"key=with=special=characters"=value`
* Keys with #: `key#hash=value`
* Keys with @: `@timestamp=2020-12-31T12:43:22.2322232Z`

Real-world examples:

* `connect=3ms delay=0ms status=500 duration=3.42s`
* `connect=68478.0e379004ms k=v`
* `path="/cart_link" host=lumberjack-store.herokuapp.com request_id=6ad70ccd-40db-477c-afce-f7e3719a886b fwd="108.30.189.26" dyno=web.1 connect=0ms service=73ms status=304 bytes=656 protocol=https`

**Comment by FungusHumungus** [comment]
I have now updated this so it uses Nom for the parsing which makes it much better at handling delimiters and escapes within them. So hopefully it will now handle all the required usecases.

Note it no longer handles the cases where wrapping delimiters are different - the only character allowed has been hardcoded to `"`.

So before it could parse things like - `<key>:((value))`. But this is no longer possible. After discussions with @binarylogic this is not a scenario we need to support right now.

File: lib/remap-functions/src/parse_key_value.rs [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+                field_split
+                    .type_def(state)
+                    .fallible_unless(value::Kind::Bytes)
+            }))
+            .merge_optional(self.separator.as_ref().map(|separator| {
+                separator
+                    .type_def(state)
+                    .fallible_unless(value::Kind::Bytes)
+            }))
+            .fallible_unless(value::Kind::Bytes)
```
Comment (FungusHumungus) [inline lib/remap-functions/src/parse_key_value.rs:96]: @JeanMertz I think this may be wrong. The function can fail even if the input is Bytes if it is not able to parse the input string correctly. So should this actually be : `.into_fallible(true)`?

I think there's a number of places in the code that do this..

Comment (JeanMertz) [reply lib/remap-functions/src/parse_key_value.rs:96]: That's correct. If we can't know at compile-time if a function can be infallible, it should always be considered fallible.

That's usually the case with `parse_*` functions, such as `parse_json`. Technically, we _could_ check if the input is a string literal, and do the parsing at compile-time, but that's such a (mostly useless) edge-case that it's not worth pursuing right now.

This also means that once we enforce infallibility at runtime, operators are required to handle error cases for all(?) parse functions.

File: lib/remap-functions/src/parse_key_value.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "field_split",
```
Comment (binarylogic) [inline lib/remap-functions/src/parse_key_value.rs:30]: What do you think about `field_delimiter` and `key_value_delimiter` instead?

File: lib/remap-functions/src/parse_key_value.rs [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        parse_separator(separator),
+        parse_key_value(field_split, separator),
+    )(input)
+}
+
+/// Parses the separator between the key/value pairs.
+/// If the separator is a space, we parse as many as we can,
+/// If it is not a space eat any whitespace before our separator as well as the separator.
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
```
Comment (FungusHumungus) [inline lib/remap-functions/src/parse_key_value.rs:117]: Any idea about this?

If I remove the lifetimes I get the following error, so it seems they are needed: 

```
error[E0759]: `separator` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
```

Comment (binarylogic) [reply lib/remap-functions/src/parse_key_value.rs:117]: cc @fanatid @pablosichert

Comment (pablosichert) [reply lib/remap-functions/src/parse_key_value.rs:117]:
```rust
fn parse_separator(separator: &str) -> impl Fn(&str) -> IResult<&str, &str> + '_
```

should do the trick.

It's mentioned in https://rust-lang.github.io/rust-clippy/master/#needless_lifetimes under _Known problems_: 

> Lifetime bounds such as impl Foo + 'a and T: 'a must be elided with the placeholder notation '_ because the fully elided notation leaves the type bound to 'static.

As to why, I also don't know exactly. Gotta read up on the `+` operator for lifetimes.

File: lib/remap-functions/src/parse_key_value.rs (hunk 2 above)
Comment (FungusHumungus) [reply lib/remap-functions/src/parse_key_value.rs:30]: Yeah, I think that's better.

File: docs/reference/remap/parse_key_value.cue [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	examples: [
+		{
+			title: "Successful match"
+			input: {
+				message: #"""
+					level=info msg="Stopping all fetchers" tag=stopping_fetchers id=ConsumerFetcherManager-1382721708341 module=kafka.consumer.ConsumerFetcherManager
+					"""#
+			}
+			source: #"""
+					. = parse_key_value(.message, field_delimiter=" ", key_value_delimiter="=")
```
Comment (JeanMertz) [inline docs/reference/remap/parse_key_value.cue:41]: Might be good to remove the options in this example, since they are the default, and it helps make clear that this function is easy to use.

It's probably best to then add a second, more complicated example, that shows how to use the two extra options.

File: lib/remap-functions/Cargo.toml [hunk 5]
```diff
@@ -128,6 +129,7 @@ parse_grok = ["grok"]
 parse_json = ["serde_json"]
 parse_regex = ["regex"]
 parse_regex_all = ["regex"]
+parse_key_value = []
```
Comment (JeanMertz) [inline lib/remap-functions/Cargo.toml:132]:
```suggestion
parse_key_value = ["nom"]
```

(you can validate this by running `cargo build --no-default-features --features parse_key_value`)

File: lib/remap-functions/src/parse_key_value.rs [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let key_value_delimiter = arguments.optional("key_value_delimiter").map(Expr::boxed);
+        let field_delimiter = arguments.optional("field_delimiter").map(Expr::boxed);
```
Comment (JeanMertz) [inline lib/remap-functions/src/parse_key_value.rs:45]: Minor nit, but you can also set the default at compile-time, to remove the extra check in the expression itself.

E.g. if `optional` returns `None`, then set it to the literal expression `"="`.

This also makes `type_def` a bit easier to implement, as you no longer need `merge_optional` but only `merge` (the outcome is the same, since literals are infallible).

File: lib/remap-functions/src/parse_key_value.rs [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+            None => "=".to_string(),
+        };
+
+        let field_delimiter = match &self.field_delimiter {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => " ".to_string(),
+        };
+
+        let (_, values) =
+            parse_line(&value, &key_value_delimiter, &field_delimiter).map_err(|e| e.to_string())?;
```
Comment (JeanMertz) [inline lib/remap-functions/src/parse_key_value.rs:78]: I didn't see any tests for failed parsing (we should add one), but I suspect we might want to add a bit more context to this error message to make it clear what happened (unless of course the error is already descriptive enough).

File: lib/remap-functions/src/parse_key_value.rs [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+
+        Ok(values.into_iter().collect::<BTreeMap<_, _>>().into())
```
Comment (JeanMertz) [inline lib/remap-functions/src/parse_key_value.rs:80]: nit: `Value` implements `FromIterator<(String, Value)>`, so you should be able to do `Ok(Value::from_iter(values))`

File: lib/remap-functions/src/parse_key_value.rs [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+            .with_constraint(value::Kind::Map)
+    }
+}
+
+/// Parse the line as a separated list of key value pairs.
+fn parse_line<'a>(
+    input: &'a str,
+    key_value_delimiter: &'a str,
+    field_delimiter: &'a str,
+) -> IResult<&'a str, Vec<(String, Value)>> {
```
Comment (JeanMertz) [inline lib/remap-functions/src/parse_key_value.rs:106]: nit: since we're only calling this function once, and we're ignoring the `&str` return value, we might as well ignore it in this function, so that the expression implementation is a bit cleaner.

File: lib/remap-functions/src/parse_key_value.rs [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+impl Expression for ParseKeyValueFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let bytes = self.value.execute(state, object)?.try_bytes()?;
+        let value = String::from_utf8_lossy(&bytes);
+
+        let key_value_delimiter = match &self.key_value_delimiter {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => "=".to_string(),
+        };
```
Comment (JeanMertz) [inline lib/remap-functions/src/parse_key_value.rs:70]: nit: since we're passing these in as string references to `parse_line`, we should be able to remove the extra `to_string()` allocation here.

File: lib/remap-functions/src/parse_key_value.rs [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+        assert_eq!(Ok(("", "noog".into())), parse_value(" ")(r#""noog""#));
+
+        // undelimited
+        assert_eq!(Ok(("", "noog".into())), parse_value(" ")("noog"));
+    }
+
+    test_type_def![
+        value_string {
+            expr: |_| ParseKeyValueFn {
+                value: Literal::from("foo").boxed(),
```
Comment (JeanMertz) [inline lib/remap-functions/src/parse_key_value.rs:247]: Nit: it doesn't matter too much, but for future reference, we have a `lit!` macro to save a few keystrokes, since we're using this often in tests (you still need to call `boxed` though).

File: lib/remap-functions/src/parse_key_value.rs [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+    ];
+
+    test_function![
+        parse_key_value => ParseKeyValue;
+
+        default {
+            args: func_args! [
+                value: r#"at=info method=GET path=/ host=myapp.herokuapp.com request_id=8601b555-6a83-4c12-8269-97c8e32cdb22 fwd="204.204.204.204" dyno=web.1 connect=1ms service=18ms status=200 bytes=13 tls_version=tls1.1 protocol=http"#,
+            ],
+            want: Ok(value!({"at": "info",
```
Comment (JeanMertz) [inline lib/remap-functions/src/parse_key_value.rs:305]: nit/remark: again, just mentioning for future reference: the map notation also allows keys without quotes (as long as they are a single identifier).

This can sometimes make maps with lots of string values easier to parse/read for others.

E.g.:

`{ "foo": "bar" }` vs `{ foo: "bar" }` (the difference is more profound with syntax highlighting)

**Review by JeanMertz — CHANGES_REQUESTED** [review]
Sweet!

Other than the missing error-case test, and the missing dependency in Cargo.toml, this looks good to go.

File: lib/remap-functions/src/parse_key_value.rs (hunk 3 above)
Comment (FungusHumungus) [reply lib/remap-functions/src/parse_key_value.rs:117]: @pablosichert Awesome, thanks! I also have some reading to do.

File: lib/remap-functions/Cargo.toml (hunk 5 above)
Comment (jszwedko) [reply lib/remap-functions/Cargo.toml:132]: :thinking: I wonder if we should add this sort of check to CI like we have for `make check-components`.

File: lib/remap-functions/src/parse_key_value.rs [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+            map(many1(tag(field_delimiter)), |_| " ")(input)
+        } else {
+            preceded(space0, tag(field_delimiter))(input)
+        }
+    }
+}
+
+/// Parse a single `key=value` tuple.
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
```
Comment (jszwedko) [inline lib/remap-functions/src/parse_key_value.rs:130]: Could we resolve this here and in the other places by using `'_` as mentioned on https://rust-lang.github.io/rust-clippy/master/#needless_lifetimes ?

**Review by jszwedko — COMMENTED** [review]
Nice, thanks @FungusHumungus !

Just a note about resolving the `needless_lifetimes` lint, and I agree with Jean's comments, but otherwise this looks good to me.

File: lib/remap-functions/src/parse_key_value.rs (hunk 13 above)
Comment (FungusHumungus) [reply lib/remap-functions/src/parse_key_value.rs]: Yes, that did work.

```rust
fn parse_value(field_delimiter: &str) -> impl Fn(&str) -> IResult<&str, Value> + '_
```

However, changes I have made to get better error messages means the lifetimes are actually needed now, so Clippy is not complaining anymore.

Comment (jszwedko) [reply lib/remap-functions/src/parse_key_value.rs]: Aha, even better :smile:

**Review by JeanMertz — APPROVED** [review]
